### ISSUE TITLE
Nextcloud notes integration (Web client)

### ISSUE THREAD

**Issue by bluefuzz** [laurent22/joplin#228]
### Operating system

- [ ] Windows
- [x] macOS
- [ ] Linux
- [ ] Android
- [ ] iOS

### Application

- [x] Desktop
- [ ] Mobile
- [ ] Terminal

With reference to issues #56 and #4 

I'm struggling to understand how this is supposed to work. If I sync Joplin with its own 'Joplin' folder on Nextcloud then it syncs fine - but all the .md files in that folder have obfuscated filenames and are unreadable by the Nextcloud notes app (or any other markdown editor). Nor does Joplin see any existing notes in the Notes folder created by QownNotes app or the Nextcloud editor ...

So is this supposed to be working or not?

**Comment by laurent22** [comment]
The files that the app creates on Nextcloud are in this format for performance reasons, they are only meant to be read and write by applications (as they are text files it is possible to view them too but as you found out it's not user-friendly).

So with this in mind you basically need to point one or more Joplin applications to this same folder on Nextcloud and they will sync with each others.

There's currently no integration with either QownNote or the Nextcloud editor.

**Comment by bluefuzz** [comment]
OK. That's a bit disappointing. I hope you consider this for a future release. 

Much of the point of using open systems like Nextcloud and open formats like markdown is to get away from closed sytems like Evernote etc. If Joplin doesn't integrate with Nextcloud's notes or even write standard markdown then that makes it essentially useless to a great many potential users ...

**Comment by sunjam** [comment]
Hey, just want to throw in my .02
    
QownNotesAPI by @pbek is very well designed and similar to Joplin, which is a good thing.  It extends simple text note taking in Nextcloud to support a Markdown editor, tag support, todo support, search functionality, evernote import.  Notes can still be indexed, shared and edited as regular md files in Nextcloud from any browser, [attachments are accessible from a folder in Nextcloud for easy export ala Joplin issue #97](https://github.com/laurent22/joplin/issues/97), you can sync everything to standard Nextcloud Notes app for organizing, and it feels easy.
    
It would totally make sense for Joplin files to be accessible in this way.  I know it can be a lot of work, but all projects would benefit immensely... Joplin would get a comprehensive server backend and a bunch of requested sharing/multi-user functionality via Nextcloud.  Plus, true note sync with Nextcloud could possibly push your mobile/terminal note apps to be the de-facto standard for all Nextcloud users, millions of them.  Nextcloud definitely needs a solid desktop/mobile/terminal note app to rule them all.  Thanks for reading.

**Comment by laurent22** [comment]
I agree some kind of web client would be nice but it would need to be a solution that doesn't involve rewriting an app from scratch. Nextcloud Notes might be a solution but I'm not completely sure it supports all the features Joplin needs. Currently we can add any feature to the Joplin format and it will still sync with any of the supported services.

If we tie it to something like Nextcloud Notes that might restrict us to just Nextcloud Notes feature set. For instance, not sure if it supports file attachments, alarms, tags, as well as the many fields that are attached to each notes (author, geo-location, user-specified creation and updated time, source application, etc.). Perhaps if Notes supported arbitrary data it might work.

Also one thing to keep in mind is whether Notes will be supported over the long time. Nextcloud is likely to last but not sure how well supported is Notes.

**Comment by sunjam** [comment]
Thanks for the thoughts.  To clarify,  I’m saying [QownnotesAPI](https://github.com/pbek/qownnotesapi), which is a [separate Nextcloud app](https://apps.nextcloud.com/apps/qownnotesapi) that plays nice with the more feature limited [Nextcloud Notes](https://apps.nextcloud.com/apps/notes), could be very useful for Joplin as a backend server option.
![5b4d1e66-12fc-4b01-81c1-013a8bc35f8f](https://user-images.githubusercontent.com/1787238/36390156-d6c0e06e-1556-11e8-96fc-c0300dd2b502.png)
[You can check out the website here... ](http://www.qownnotes.org) it currently supports the cross platform Qownnotes desktop app for Mac, Windows and Linux + plays nice with Evernote imported files.

**Comment by sunjam** [comment]
[The markdown files in Joplin also do not work with the default Nextcloud Markdown Editor due to metadata conflicts from Joplin](https://github.com/icewind1991/files_markdown)

**Comment by sunjam** [comment]
[Want to add a friendly link to the NextNotes fork, which is a current unmaintained WYSIWYG markdown related notes server for Evernote data on Nextcloud.](https://github.com/janis91/nextnotes).. Sooo glad to see this issue re-opened as a possible interest in maintaining a universal open source standard on Joplin.

**Comment by e-alfred** [comment]
Nextnote would be another alternative which currently misses desktop/mobile editors:

https://github.com/brantje/nextnote

**Comment by sunjam** [comment]
> Nextnote would be another actively maintained alternative

https://t.me/nextnote is a direct link to @brantje dev Telegram chat for the project.

**Comment by milankni** [comment]
Until a web client exists, I tend to look up the particular note .md file with the help of fulltext search (ElasticSearch) in Nextcloud and then edit the file (leaving the metadata untouched).
Links are broken, obviously, but editing larger text is easier than on mobile phone.

**Comment by sunjam** [comment]
@milankni that is exactly what this feature request hopes to improve.  👍

**Comment by krueuw** [comment]
The QOwnNotes project has similar goals to Joplin. It has a desktop operating system app that is incompatible with Joplin. Please work together, talk about an API which supports both projects and makes something great compatible! Any necessary compromises are worth the price - mobile stable access and sophisticated desktop features on the notes on Owncloud and NextCloud.

**Comment by bernd-wechner** [comment]
I would just add a vote of support for integration with a NextCloud app, ideally the default NextCloud Notes app: https://apps.nextcloud.com/apps/notes

That would take Joplin to a whole new level of cool for some of us I'm sure!

**Comment by foxmask** [comment]
I made [something like that](https://discourse.joplin.cozic.net/t/jong-importer-creating-note-from-your-favorite-cloud-service/486) this weekend . Could be helpful ?

**Comment by foxmask** [comment]
In the meantime I started [this joplin-web](https://github.com/foxmask/joplin-web/) project this weekend :)

**Comment by bernd-wechner** [comment]
> Does it have to integrate with the Nextcloud app at first?

No, of course not.  But for those of us with nextcloud servers (and we apparently number in the 25 million already ;-)

https://nextcloud.com/blog/japan-to-add-millions-of-new-nodes-to-federated-nextcloud-network/

There is already a "Notes" app. And we would love to see cooperation between developers of related apps rather than divergence and can't see why the NextCloud Notes app can't be the database store for Joplin rather than WebDav .md files and an interface for editing them. That is all. 

In short, that would rock, and not be a majorly hard to do at a gues.

**Comment by i-need-to-tell-you-something** [comment]
To translate many of these thoughts to practical terms:

There is end-user desire for either of:
1. QOwnNotes to switch their backend to be Joplin-compatible.
2. NextNote to switch their backend to be Joplin-compatible.
3. A new Nextcloud app with Joplin-compatible backend.

<Skip this rant> Just in case some might not realize:
Option 3 entails a maintenance cost every month to make the Nextcloud-app conform to changes in the rest of Joplin. Some person would have to start giving away their free time for this all the time. This could be also done by the person(s) currently working on Joplin, but this would obviously mean that time spent on this would be time not spent on making the rest of Joplin experience. And they can't just spend more time on Joplin, since that would mean less personal life or less money to pay for rent and food.
Options 1 and 2 would incur a huge initial cost on switching from current backend to a Joplin compatible one. And then the maintenance cost as well, but they might be paying a larger or smaller cost if they were to keep doing their own version of the backend. Unless they are being paid for it, they might not have great enthusiasm to spend all this time for this conversion.

All of this might sound pessimistic, but there are a lot of generous people doing great work for us all every day for no significant benefit of their own. I'm just telling some people to not take all of this work for granted. There isn't a magical slave race of elves/goblins doing this work in some sweatshops for us. So sometimes we should skip a candybar or eating out or a toy we don't *really* need, to give some money to these people despite the effort.</Skip this rant>

**Comment by sunjam** [comment]
My understanding of the thread is Joplin could make their notes compatible with standard markdown or release documentation of the JEX format for others to follow.

- Standard Markdown compatibility would allow notes to work with Nextcloud Notes and Qownnotes as-is, while missing out on the nicer features found in Jex.
- [Jex format documentation](https://github.com/laurent22/joplin/issues/408) would allow others to support Joplin format, which [Jan of Nextcloud mentioned being interested in](https://github.com/laurent22/joplin/issues/408#issuecomment-442640408) as an open standard once documented.

Definitely all community, volunteer work.  Releasing some nice documentation of JEX would be awesome!  The road to a full webui is complex, and these are all excellent projects with different devs.

[Send donations here via Paypal or Patreon to support Joplin](https://github.com/laurent22/joplin/blob/master/readme/donate.md)

**Comment by JokerGermany** [comment]
Is this Issue about creating a NextCloud Webapp or shall i create a new issue? for this one?
Until the Nextcloud App i will use Carnet instead:
https://github.com/PhieF/CarnetDocumentation/

**Comment by vbatts** [comment]
Carnet is not even close to a replacement of joplin.

I currently use joplin cli, linux desktop and android mobile, all backended by my NextCloud. If it were possible to _also_ view and edit notes via the nextcloud web UI this would be _amazing_ (i.e. I am using a loner/work laptop where all I can use is a browser; I'm doing research at a kiosk style library computer (browser only)).

**Comment by foxmask** [comment]
[in case that could be a possibility for the nextcould users](https://discourse.joplinapp.org/t/joplin-web-web-application-companion-for-joplin/555/24)

**Comment by freekvh** [comment]
I'd also like to add my 2ct, I thought Joplin was a markdown editor so I point it at the Notes folder on my Nextcloud instance but I don't see my existing notes. Nextcloud has a new (very nice) Markdown editor build in and that does nicely play with Joplins files. So why not the other way around too?

**Comment by bernd-wechner** [comment]
Fully agreed. It would simply be superb to see Joplin Notes on the Nextcloud Notes app! Something to dream of!

**Comment by PackElend** [comment]
the idea is to make the new API or a port of the web companion a project in the GSoC season next year if Joplin gets accepted.
We are happy about any support, join us on
[Joplin is applying for Google Summer of Code 2020](https://discourse.joplinapp.org/t/joplin-is-applying-for-google-summer-of-code-2020/4324/1)

**Comment by laurent22** [comment]
Was tried but won't be done due to the poor quality of the Nextcloud API
